### ISSUE TITLE
[Prover] Error handling borrow_mut_global-like native functions

### ISSUE THREAD

**Issue by awelc** [move-language/move#660]
With different underlying storage models for different Move flavors, one may have to implement their own version of operators accessing global storage. Currently, without baking it into prover itself, this is only possible via native functions, but what is supported/possible right now is not always sufficient. In particular, the borrow checker complains when trying to prove the following module:
```rust
module broken_native::broken_native {

    struct SomeStruct has key, drop {
        val: u64
    }

    native fun just_like_borrow_mut_global<T: key>(a: address): &mut T;

    fun foo(a: address) {
        let r = just_like_borrow_mut_global<SomeStruct>(a);
        r.val = 42;
    }
}
```

The error generated looks as follows:
```
thread 'main' panicked at 'inconsistent borrow information: undefined output: broken_native::just_like_borrow_mut_global
', language/move-prover/bytecode/src/borrow_analysis.rs:324:17
```

I am opening this issue to start a discussion on how this could be supported in the prover to facilitate different storage models. @emmazzz made an observation that the native function in question attempts to do something similar to what the `borrow_global_mut` operation does and perhaps the same mechanism that's used to handle the latter could also be used to handle the former (@wrwg, @stephenreborn, @junkil-park?)

**Comment by meng-xu-cs** [comment]
The way I am thinking is that we can use the `pragma intrinsic` scheme to annotate that this `just_like_borrow_mut_global` funtion is essentially an intrinsic function which behaves just like `borrow_mut_global`.

Syntatically:

```
native fun just_like_borrow_mut_global<T: key>(a: address): &mut T;

spec just_like_borrow_mut_global {
    pragma intrinsic = borrow_global_mut;
}
```

We already use the similar approach to support variable implementation of the `Map` type, as shown in the `Table.spec.move` file.
```
spec Table {
    pragma intrinsic = map;
}
```

**Comment by awelc** [comment]
I was erring on the side of caution and wanted to be able to uniquely identify a function in question to avoid a situation where somewhere in someone's code we'd have a function with the intrinsic pragma and the same name.

Also, I am not sure how to easily (or easier than what I implemented) to check if a given function has intrinsic pragma. We have `is_intrinsic_of` but that's for structs rather than for functions. I am sure it could be added but I wasn't sure if such addition would be warranted by this single use case.

**Comment by meng-xu-cs** [comment]
> avoid a situation where somewhere in someone's code we'd have a function with the intrinsic pragma and the same name.

I am not entirely sure that I understand this one.
- I don't think we can have two move functions having the same name (well same `account::module::function` name)
- It is perfectly fine to have two functions being intrinsic to the same thing (like having `Table` and `SmallMap` both being intrinsic to `map`). It is perfectly legal to have to function, e.g., `i_am_borrow_global_mut` and `i_am_also_borrow_global_mut` both being `pragma intrinsic = borrow_global_mut;`.

**Comment by awelc** [comment]
I am sorry if I wasn't clear enough. Also perhaps I misunderstood the meaning of the comment, which in my mind was about how we identify the native function we want to treat specially - the suggestion being to use the `intrinsic` pragma rather than explicitly specifying module address, module name, and function name.

What I was trying to say is that we need to uniquely identify a function which I am not sure how the `intrinsic` pragma helps with. Say, we have a native framework function that we want to treat specially `i_am_borrow_global_mut`, which is in module `framework` at address `0x2`. Even if we mark it as `intrinsic`, we still have to check if this function is defined in module `framework` at address `0x2`, otherwise we may have a developer accidentally name a function in their module `i_am_borrow_global_mut` and also mark it as `intrinsic` and have the Prover (try to) treat is specially even though it shouldn't.

In other words, I am not sure what advantage "overloading" the meaning of `intrinsic` to mark these special functions has over what's already implemented.

**Comment by awelc** [comment]
> In fact, what I try to say is that: what is given to the prover via the intrinsic function is going to be (or more pessimistically) implicitly trusted...

This makes sense to me, as this is close what the `intrinsic`'s original intent is according to the docs. There are still some things that I am missing here it seems...

> How to tell the prover that i_am_borrow_global_mut is essentially borrow_global_mut? ==> the solution is to overload the intrinsic pragma

Are you suggesting that marking a function intrinsic is enough for the the prover to treat it as `borrow_global_mut`? Surely, you also need to tell the prover the name of this function - `intrinsic` pragma can be slapped on any (even native) function that may have nothing to do with borrow semantics and we don't want to treat all of them in the same way. You also need to tell the prover which operations implemented in Boogie it has to use to model reading/writing to storage (for vector and `Table` these are hard-coded, e.g., `GetTable`, `UpdateTable`). In other words, I think you need to pass "custom" parameters (currently done via `BorrowNative` struct) to the prover to get this to work, and I just don't see how overloading `intrinsic` helps with that. 

I do agree, however, that this "borrow mut" behavior will only apply to native functions, which can only be defined by the adapter writers. This would allow simplifying the check for the functions that need special treatment. We could simplify custom parameter to only contain function name and names of the Boogi operations:
```
pub struct BorrowNative {
    /// Method's name
    pub name: String,
    /// Name of a read operation for a given type (in Boogie)
    pub read_op: String,
    /// Name of a Write operation for a given type (in Boogie)
    pub write_op: String,
}
```
During the actual check, we'd check if the function is native and if the name matches.

Please note that I am not trying to be explicitly difficult here. I am happy to make changes, but at this point I just don't know what the changes should be... I still don't see how adding `intrinsic` annotation check helps here since (I think) we still need additional checks (and the additional data passed in). 

It would be really helpful if you could provide comments in specific code sections explaining how you'd like to see things modified for everything to work.

**Comment by meng-xu-cs** [comment]
One confusion from me is that, I don't see this `struct BorrowNative` being in the code base. Does it live in a specific PR or branch or repo? Without seeing it, it is extremely difficult to guess out your solution.

> You also need to tell the prover which operations implemented in Boogie it has to use to model reading/writing to storage

I am completely lost. Modeling reading/writing to storage is sth the prover does (and understands), not what Boogie does. The semantics of storage show up in `move-model` (i.e., stackless bytecode and expressions), several passes of the bytecode transformation pipeline (e.g., borrow analysis), translation to Boogie as well as in prelude. I have no idea what you are referring to here.

> I think you need to pass "custom" parameters (currently done via BorrowNative struct) to the prover to get this to work

Again, I don't understand what "custom" mean here. Any examples?

> Please note that I am not trying to be explicitly difficult here. I am happy to make changes, but at this point I just don't know what the changes should be...

Sorry, I don't even know what your original solution is, therefore I cannot make any suggestion. I am not asking for any changes and I am not sticking to the `intrinsic` design. I am only saying that this is what comes out of my mind. If you already have a solution which works, I am more than happy to hear it out.

**Comment by meng-xu-cs** [comment]
And one piece of information you might not have realized is that: **not all native functions are equal** in the prover's point of view.

For regular native function (e.g., native functions with `bcs` or crypto-operations), they can be modeled directly in Boogie by augmenting the `prelude.bpl`.

Native functions that have an impact on the borrow semantics of a program *and/or* global states (which are not a built-in concept in Boogie) needs to be recognized way earlier in `move-model` and the bytecode transformation pipeline. Unfortunately, in this particular case, `just_like_borrow_mut_global` is in the latter category --- a function that the prover must be directly informed.

**Comment by awelc** [comment]
OMG... I am so sorry!!! Your initial comment was so on point that I completely did not realize we are commenting on the issue rather than the PR. It's my fault entirely and I am sorry for wasting your time... Here's the PR I thought we were discussing and the code I was referring to: https://github.com/move-language/move/pull/679
